### PR TITLE
EP Merge: Add claim contention metrics

### LINKED ISSUE #3065: Implement metrics for EP Merge for single-contention new/cfi claims.
## Summary
With the addition of processing Merge requests with single contention new claims, it would be useful to be able to know the volume of new claims vs claims for increase. 

## Acceptance Criteria
* Metrics are defined
* Metrics are implemented
* Unit tests are added

### PR DESCRIPTION
**Pull request by dfitchett**
## What was the problem?

EP Merge app does not publish metrics on the number or type of contentions on a pending EP or EP400.  The only metric related to this is the number of contentions that were merged.

Associated tickets or Slack threads:
- #3065

## How does this fix it?

This adds metrics that capture the following: 
* `job.merged_contentions` - previously captured as `job.new_contentions`
* `job.merged_new_contentions` - # of new contentions on the merged claim
* `job.merged_cfi_contentions` - # of cfi contentions on the merged claim
* `pending_ep.new_contentions` - # of new contentions on the pending claim
* `pending_ep.cfi_contentions` - # of cfi contentions on the pending claim
* `ep400.new_contentions` - # of new contentions on the ep400 claim
* `ep400.cfi_contentions` - # of cfi contentions on the ep400 claim

## How to test this PR
- pull code
- Unit Test:
  - from /domain-ee/ee-ep-merge-app run pytest
- Integration Tests:
  - run [base platform](https://github.com/department-of-veterans-affairs/abd-vro/wiki/Docker-Compose#platform-base)
  - from /domain-ee/ee-ep-merge-app run pytest ./integration
- End to End tests:
  - run this [CI](https://github.com/department-of-veterans-affairs/abd-vro/actions/workflows/ee-ep-merge-end-to-end.yml) action to verify successful integration test
  - run end to end tests locally:
    - `export BIP_CLAIM_URL=mock-bip-claims-api:20300`
    - run bgs and bip services and mocks:
      - `COMPOSE_PROFILES=“all” ./gradlew dockerComposeUp`
      - `COMPOSE_PROFILES="bip,bgs" ./gradlew -p mocks :dockerComposeUp`
    - `./gradlew :domain-ee:ee-ep-merge-app:endToEndTest`

### CODE REVIEW

File: domain-ee/ee-ep-merge-app/src/python_src/service/ep_merge_machine.py [hunk 1]
```diff
@@ -165,6 +166,10 @@ class EpMergeMachine(StateMachine):  # type: ignore[misc]
     def __init__(self, merge_job: MergeJob, main_event: Workflow = Workflow.PROCESS):
         self.job = merge_job
         self.main_event = main_event
+        self.cancellation_reason: str | None = None
+        self.original_tsoj: str | None = None
+        self.ep_metrics: list[DistributionMetric] = []
+        self.skipped_merge: bool = True
```
Comment (dfitchett) [inline domain-ee/ee-ep-merge-app/src/python_src/service/ep_merge_machine.py:172]: Moving these to instance variables prevents issues with potential concurrent modification of cancellation_reason and original TSOJ.

File: domain-ee/ee-ep-merge-app/src/python_src/service/ep_merge_machine.py [hunk 2]
```diff
@@ -230,6 +235,16 @@ def on_get_pending_contentions(self, event):
             response_type=get_contentions.Response,
             expected_statuses=expected_statuses,
         )
+        if response and response.status_code in expected_statuses:
+            contentions = response.contentions or []
+            num_pending_ep_new_contentions = sum(1 for c in contentions if c.contention_type_code == 'NEW')
+            num_pending_ep_cfi_contentions = sum(1 for c in contentions if c.contention_type_code == 'INCREASE')
```
Comment (nanotone) [inline domain-ee/ee-ep-merge-app/src/python_src/service/ep_merge_machine.py]: Booleans convert to 0 and 1, so these can be rephrased like

```python
sum(c.contention_type_code == 'NEW' for c in contentions)
```

which is not a strong convention, but I think it reads fine if you think of `sum` like "count."

File: domain-ee/ee-ep-merge-app/src/python_src/util/metric_logger.py [hunk 3]
```diff
@@ -31,66 +35,98 @@
...
+        self.name = name
+        self.value = value
+
+    def __eq__(self, other):
+        if not isinstance(other, self.__class__):
+            return False
+        return self.name == other.name and self.value == other.value
+
+    def __hash__(self):
+        return hash((self.name, self.value))
```
Comment (nanotone) [inline domain-ee/ee-ep-merge-app/src/python_src/util/metric_logger.py]: It doesn't look like you need any nuanced mutability stuff with name and value, so you could use Python's included battery [typing.NamedTuple](https://docs.python.org/3/library/typing.html#typing.NamedTuple) for this.

File: domain-ee/ee-ep-merge-app/src/python_src/util/metric_logger.py [hunk 4]
```diff
@@ -31,66 +35,98 @@
...
+    def __init__(self, name: str, value: float = 1):
+        super().__init__(name, value)
+
+    def __repr__(self):
+        return f'CountMetric(name={self.name}, value={self.value})'
+
+
+class DistributionMetric(Metric):
+    def __repr__(self):
+        return f'DistributionMetric(name={self.name}, value={self.value})'
```
Comment (nanotone) [inline domain-ee/ee-ep-merge-app/src/python_src/util/metric_logger.py]: And in fact, subclassing NamedTuple gives you, for free, this exact `repr` implementation 😉

File: domain-ee/ee-ep-merge-app/src/python_src/util/metric_logger.py [hunk 5]
```diff
@@ -31,66 +35,98 @@
...
+    Adds value to a count metric with by the name '{APP_PREFIX}.{metric.name.strip(".").lower()}'
+    :param metrics: list of CountMetric objects
+    """
+    series = [
+        MetricSeries(
+            metric=f'{APP_PREFIX}.{metric.name.strip(".").lower()}',
+            type=MetricIntakeType.COUNT,
+            points=[
+                MetricPoint(
+                    timestamp=int(datetime.now().timestamp()),
```
Comment (nanotone) [inline domain-ee/ee-ep-merge-app/src/python_src/util/metric_logger.py]: I missed this the first "time" around, but if you don't need fancy human stuff you can just do `time.time()` to get Unix time.

**Review by nanotone — APPROVED** [review]
Left a handful of style suggestions, but none that are blocking.

Question though: does VRO Datadog actually let you do histogram/distribution widgets with these distribution metrics?

File: domain-ee/ee-ep-merge-app/src/python_src/util/metric_logger.py (hunk 5 above)
Comment (dfitchett) [reply domain-ee/ee-ep-merge-app/src/python_src/util/metric_logger.py]: Ah the examples on datadogs site for the API show that same line I have, but `int(time.time())` should work just fine too!

**Comment by dfitchett** [comment]
> Question though: does VRO Datadog actually let you do histogram/distribution widgets with these distribution metrics?

Having it as a histogram/distribution metric isn't adding a whole lot of value currently as we only have a single instance running and such few jobs that actually get executed. Basically, its just better than count / gauge metrics if the metric needs to know individual values. If more jobs were running concurrently, then we could get the case where during the flush interval `count` would add all of the values of a metric up and `guage` would only return the last value. TBH I'm not sure based on our traffic if its really worth it. We could probably change all our metrics to `count` since we have such little traffic.

**Pull request merged**